torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -311,15 +337,62 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   }
 
   /// Adds a type-erased `AnyModule` to the `Sequential`.
+  ///
+  /// See `add_to_modules(AnyModule any_module)` for implementation detail.
   void push_back(AnyModule any_module) {
-    modules_.push_back(std::move(any_module));
-    const auto index = modules_.size() - 1;
+    auto index = add_to_modules(any_module);
     register_module(std::to_string(index), modules_[index].ptr());
   }
 
   /// The base case, when the list of modules is empty.
   void push_back() {}
 
+  /// Adds a new (boxed) `Module` to the `Sequential` container, and returns
+  /// its index in `modules_`.
+  template <typename ModuleType>
+  size_t add_to_modules(std::shared_ptr<ModuleType> module_ptr) {
+    // Nesting Sequential doesn't work because `forward()`'s return type is
+    // templatized, so it'll give a nasty compiler error.
+    static_assert(
+        !std::is_same<SequentialImpl, ModuleType>::value,
+        "Sequential is not nestable");
+    static_assert(
+        torch::detail::is_module<ModuleType>::value,
+        "Can only add objects derived from nn::Module to Sequential");
+    static_assert(
+        torch::detail::has_forward<ModuleType>::value,
+        "Can only add modules with a forward() method to Sequential");
+    return add_to_modules(AnyModule(std::move(module_ptr)));
+  }
+
+  /// Adds a new `Module` to the `Sequential` container, moving or copying it
+  /// into a `shared_ptr` internally, and returns its index in `modules_`.
+  //// This method allows passing value types,
+  /// and letting the container deal with the boxing. This means you can write
+  /// `Sequential(Module(3, 4))` instead of
+  /// `Sequential(std::make_shared<Module>(3, 4))`.
+  template <typename M, typename = torch::detail::enable_if_module_t<M>>
+  size_t add_to_modules(M&& module) {
+    // Need to get rid of any reference components for make_unique.
+    using Type = typename std::remove_reference<M>::type;
```

Comment:
I also don't quite get the point of `add_to_modules` (are you always properly moving/forwarding parameters around)?  Why don't the "final" non-string versions of `push_back` just call the string versions?